SplitNCigarReads fails with NoClassDefFoundError

SplitNCigarReads is failing due to incompatibilities between htsjdk's version of snappy and Spark's version.

Temporary solution is to add system property 'disable.snappy' to force htsjdk to fallback to pure java.  Longer term solution likely involves patches to htsjdk and possibly snappy itself.

```
 ./gatk-launch SplitNCigarReads -I src/test/resources/large/CEUTrio.HiSeq.WGS.b37.NA12878.20.21.bam -O out.bam -R src/test/resources/large/human_g1k_v37.20.21.fasta

Running:
    /Users/louisb/Workspace/gatk/build/install/gatk/bin/gatk SplitNCigarReads -I src/test/resources/large/CEUTrio.HiSeq.WGS.b37.NA12878.20.21.bam -O out.bam -R src/test/resources/large/human_g1k_v37.20.21.fasta
15:31:00.516 INFO  IntelGKLUtils - Trying to load Intel GKL library from:
	jar:file:/Users/louisb/Workspace/gatk/build/install/gatk/lib/gkl-0.1.2.jar!/com/intel/gkl/native/libIntelGKL.dylib
15:31:00.552 INFO  IntelGKLUtils - Intel GKL library loaded from classpath.
[July 20, 2016 3:31:00 PM EDT] org.broadinstitute.hellbender.tools.walkers.rnaseq.SplitNCigarReads  --output out.bam --input src/test/resources/large/CEUTrio.HiSeq.WGS.b37.NA12878.20.21.bam --reference src/test/resources/large/human_g1k_v37.20.21.fasta  --refactor_NDN_cigar_string false --maxReadsInMemory 150000 --maxMismatchesInOverhang 1 --maxBasesInOverhang 40 --doNotFixOverhangs false --disable_all_read_filters false --interval_set_rule UNION --interval_padding 0 --readValidationStringency SILENT --secondsBetweenProgressUpdates 10.0 --disableSequenceDictionaryValidation false --createOutputBamIndex true --createOutputBamMD5 false --addOutputSAMProgramRecord true --help false --version false --verbosity INFO --QUIET false --use_jdk_deflater false
[July 20, 2016 3:31:00 PM EDT] Executing as louisb@wm1b0-8ab on Mac OS X 10.10.5 x86_64; Java HotSpot(TM) 64-Bit Server VM 1.8.0_91-b14; Version: Version:4.alpha.1-217-g3ff51ed-SNAPSHOT
15:31:00.557 INFO  SplitNCigarReads - Defaults.BUFFER_SIZE : 131072
15:31:00.557 INFO  SplitNCigarReads - Defaults.COMPRESSION_LEVEL : 1
15:31:00.557 INFO  SplitNCigarReads - Defaults.CREATE_INDEX : false
15:31:00.557 INFO  SplitNCigarReads - Defaults.CREATE_MD5 : false
15:31:00.557 INFO  SplitNCigarReads - Defaults.CUSTOM_READER_FACTORY :
15:31:00.557 INFO  SplitNCigarReads - Defaults.EBI_REFERENCE_SERVICE_URL_MASK : http://www.ebi.ac.uk/ena/cram/md5/%s
15:31:00.557 INFO  SplitNCigarReads - Defaults.NON_ZERO_BUFFER_SIZE : 131072
15:31:00.557 INFO  SplitNCigarReads - Defaults.REFERENCE_FASTA : null
15:31:00.557 INFO  SplitNCigarReads - Defaults.SAM_FLAG_FIELD_FORMAT : DECIMAL
15:31:00.557 INFO  SplitNCigarReads - Defaults.USE_ASYNC_IO_READ_FOR_SAMTOOLS : false
15:31:00.557 INFO  SplitNCigarReads - Defaults.USE_ASYNC_IO_WRITE_FOR_SAMTOOLS : true
15:31:00.557 INFO  SplitNCigarReads - Defaults.USE_ASYNC_IO_WRITE_FOR_TRIBBLE : false
15:31:00.557 INFO  SplitNCigarReads - Defaults.USE_CRAM_REF_DOWNLOAD : false
15:31:00.558 INFO  SplitNCigarReads - Deflater IntelDeflater
15:31:00.558 INFO  SplitNCigarReads - Initializing engine
15:31:00.659 INFO  SplitNCigarReads - Done initializing engine
15:31:00.679 INFO  ProgressMeter - Starting traversal
15:31:00.679 INFO  ProgressMeter -        Current Locus  Elapsed Minutes     Records Processed   Records/Minute
15:31:05.088 INFO  SplitNCigarReads - Shutting down engine
[July 20, 2016 3:31:05 PM EDT] org.broadinstitute.hellbender.tools.walkers.rnaseq.SplitNCigarReads done. Elapsed time: 0.08 minutes.
Runtime.totalMemory()=1011875840
Exception in thread "main" java.lang.NoClassDefFoundError: org/xerial/snappy/LoadSnappy
	at htsjdk.samtools.util.SnappyLoader.<init>(SnappyLoader.java:86)
	at htsjdk.samtools.util.SnappyLoader.<init>(SnappyLoader.java:52)
	at htsjdk.samtools.util.TempStreamFactory.getSnappyLoader(TempStreamFactory.java:42)
	at htsjdk.samtools.util.TempStreamFactory.wrapTempOutputStream(TempStreamFactory.java:74)
	at htsjdk.samtools.util.SortingCollection.spillToDisk(SortingCollection.java:223)
	at htsjdk.samtools.util.SortingCollection.add(SortingCollection.java:166)
	at htsjdk.samtools.SAMFileWriterImpl.addAlignment(SAMFileWriterImpl.java:192)
	at htsjdk.samtools.AsyncSAMFileWriter.synchronouslyWrite(AsyncSAMFileWriter.java:36)
	at htsjdk.samtools.AsyncSAMFileWriter.synchronouslyWrite(AsyncSAMFileWriter.java:16)
	at htsjdk.samtools.util.AbstractAsyncWriter$WriterRunnable.run(AbstractAsyncWriter.java:117)
	at java.lang.Thread.run(Thread.java:745)
Caused by: java.lang.ClassNotFoundException: org.xerial.snappy.LoadSnappy
	at java.net.URLClassLoader.findClass(URLClassLoader.java:381)
	at java.lang.ClassLoader.loadClass(ClassLoader.java:424)
	at sun.misc.Launcher$AppClassLoader.loadClass(Launcher.java:331)
	at java.lang.ClassLoader.loadClass(ClassLoader.java:357)
	... 11 more
```

--- a/build.gradle
+++ b/build.gradle
@@ -26,7 +26,7 @@ import de.undercouch.gradle.tasks.download.Download
 mainClassName = "org.broadinstitute.hellbender.Main"
 
 //Note: the test suite must use the same defaults. If you change system properties in this list you must also update the one in the test task
-applicationDefaultJvmArgs = ["-Dsamjdk.use_async_io_read_samtools=false","-Dsamjdk.use_async_io_write_samtools=true", "-Dsamjdk.use_async_io_write_tribble=false", "-Dsamjdk.compression_level=1"]
+applicationDefaultJvmArgs = ["-Dsamjdk.use_async_io_read_samtools=false","-Dsamjdk.use_async_io_write_samtools=true", "-Dsamjdk.use_async_io_write_tribble=false", "-Dsamjdk.compression_level=1", "-Dsnappy.disable=true"]
 
 //Delete the windows script - we never test on Windows so let's not pretend it works
 startScripts {
@@ -240,6 +240,7 @@ test {
     systemProperty "samjdk.use_async_io_write_samtools", "true"
     systemProperty "samjdk.use_async_io_write_tribble", "false"
     systemProperty "samjdk.compression_level", "1"
+    systemProperty "snappy.disable", "true"
     systemProperty "gatk.spark.debug", System.getProperty("gatk.spark.debug")
 
     // set heap size for the test JVM(s)

--- a/src/test/java/org/broadinstitute/hellbender/tools/walkers/rnaseq/SplitNCigarReadsIntegrationTest.java
+++ b/src/test/java/org/broadinstitute/hellbender/tools/walkers/rnaseq/SplitNCigarReadsIntegrationTest.java
@@ -1,9 +1,11 @@
 package org.broadinstitute.hellbender.tools.walkers.rnaseq;
 
 import org.broadinstitute.hellbender.CommandLineProgramTest;
+import org.broadinstitute.hellbender.utils.test.ArgumentsBuilder;
 import org.broadinstitute.hellbender.utils.test.IntegrationTestSpec;
 import org.testng.annotations.Test;
 
+import java.io.File;
 import java.util.Arrays;
 
 /**
@@ -68,5 +70,15 @@ public final class SplitNCigarReadsIntegrationTest extends CommandLineProgramTes
                 "-R" + b37_reference_20_21 + " -I " + largeFileTestDir + "NA12878.RNAseq.bam -O %s -L 20:2444518-2454410",
                 Arrays.asList(largeFileTestDir + "expected.NA12878.RNAseq.splitNcigarReads.subSequenceTest.bam")); //results created using gatk3.5
         spec.executeTest("regression test for unmapped and unpaired reads", this);
+    }
+
+    @Test //regression test for https://github.com/broadinstitute/gatk/issues/2026
+    public void testLargeFileThatForcesSnappyUsage(){
+        final ArgumentsBuilder args = new ArgumentsBuilder()
+                .addReference(new File(b37_reference_20_21))
+                .addInput(new File(largeFileTestDir, "CEUTrio.HiSeq.WGS.b37.NA12878.20.21.bam"))
+                .addOutput(createTempFile("largeSplitNCigarReadsTest",".bam"));
+        //Just make sure this doesn't fail with NoClassDefFoundError
+        runCommandLine(args);
     }
 }

